### PR TITLE
AC_Avoidance: fixes to work with SCurves

### PR DESCRIPTION
**Pull request by rmackay9**
This PR builds on the SCurves PR https://github.com/ArduPilot/ardupilot/pull/15896 to fixup Dijkstra's and BendyRuler's performance when using SCurves.

- Location::linear_interpolate_alt to allow setting this's alt to a linear interpolation of the lat from and origin and destination.  This is used with BendyRuler which only works horizontally.
- BendyRuler and PathPlanner are enhanced so the caller knows which path planner was used (previously the caller didn't need to know, they would just pass whatever destination was returned into AC_WPNav)
- WPNav_OA is enhanced so that it handles a destination from BendyRuler differently than a destination from Dikstra's.  Dijkstra's destinations continue to be passed into AC_WPNav (and SCurves) while BendyRuler's results go into the position controller

There are also some changes to formatting and spelling.

This has been tested in SITL but the following issues remain.  It is possible we could resolve these in a follow-up PR.

- BendyRuler can return a position outside the fence (this is an existing issue) meaning the vehicle may stray outside the fence
- BendyRuler's LOOKAHEAD distance can be shorter than the vehicle's stopping distance (which comes from WPNAV_ACCEL and WPNAV_SPEED)

### CODE REVIEW

**Review by lthall — APPROVED** [review]
All good!

**Comment by rishabsingh3003** [comment]
Things are looking good! Everything working same as Master on the simulators. Will fly it soon :)

**Comment by peterbarker** [comment]
This one needs a rebase.

**Comment by rmackay9** [comment]
@peterbarker, it'll need a lot more than that actually .. don't worry about this one for now, I've got it on my to-do list to update it once Leonard's finalised his changes to Guided mode.

**Comment by rishabsingh3003** [comment]
https://discuss.ardupilot.org/t/i-have-a-problem-with-object-avoidance-with-bendy-ruler-4-1-0-beta-3-firmware/73361
Some more evidence apart from my testing

File: libraries/AC_WPNav/AC_WPNav_OA.cpp [hunk 1]
```diff
@@ -107,26 +110,94 @@ bool AC_WPNav_OA::update_wpnav()
...
+                // Note: do not update yaw or yaw rate as we do for BendyRuler horizontal
+
+                // return success without calling parent AC_WPNav
+                return true;
             }
-            break;
-        }
+
+            } // case AP_OAPathPlanner::OA_SUCCESS
+        } // switch (oa_retstate)
```
Comment (rishabsingh3003) [inline libraries/AC_WPNav/AC_WPNav_OA.cpp:200]: Some left over commented out code

**Review by rishabsingh3003 — CHANGES_REQUESTED** [review]

File: libraries/AC_WPNav/AC_WPNav_OA.cpp (hunk 1 above)
Comment (rmackay9) [reply libraries/AC_WPNav/AC_WPNav_OA.cpp]: removed, thanks.

**Review by rishabsingh3003 — APPROVED** [review]
Tested this quite a bit
https://www.youtube.com/watch?v=d1EYWhcK2OA

It looks okay to me, but the reactions are rather twitchy at times (some of the twitches were there before also). This can be seen on real vehicles too.. It's a step forward because as of now, scurves totally breaks OA, but there is scope for improvement I guess..

**Comment by rmackay9** [comment]
@rishabsingh3003 thanks for the testing!

Merged!

**Comment by hendjoshsr71** [comment]
It also fixed the issue with vertical bendy climbing between waypoints when terrain following is on.

**Pull request merged**